src/lua/xml.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "lua/xml.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <lua.h>
+#include <lauxlib.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "lua/utils.h"
+
+int
+luaT_xml_decode(struct lua_State *L)
+{
+	/* The input string must be at the top of the stack. */
+	int top = lua_gettop(L);
+	if (top < 1 || lua_type(L, top) != LUA_TSTRING)
+		return luaL_error(L, "expected string");
+	const char *s = lua_tostring(L, top);
+	assert(s != NULL);
+	/* Push the document table. */
+	lua_newtable(L);
+	enum {
+		ELEM,
+		TAG,
+		START_TAG,
+		END_TAG,
+		SPACE_AFTER_END_TAG,
+		ATTR,
+		ATTR_NAME,
+		ATTR_VALUE_SEP,
+		ATTR_VALUE_BEGIN,
+		ATTR_VALUE,
+		ATTR_VALUE_END,
+		ELEM_END,
+		DOC_END,
+	} state = ELEM;
+	const char *tag_name = NULL;
+	const char *attr_name = NULL;
+	const char *attr_value = NULL;
+	/*
+	 * Position of the current character in the input string recorded
+	 * as line and column numbers. Used for error reporting.
+	 */
+	int line = 1;
+	int column = 1;
+#define ERROR(msg) \
+	luaL_error(L, "XML decode error at line %d, column %d: %s", \
+		   line, column, (msg))
+	while (*s != '\0') {
+		/*
+		 * State machine:
+		 *  - 'break' skips the current character by leaving
+		 *    the switch-case.
+		 *  - 'continue' rechecks the state without skipping
+		 *    the current character.
+		 */
+		switch (state) {
+		/*
+		 * Expected an element. Currently, only tags are supported
+		 * (<section> or </section>) while values enclosed in tags are
+		 * considered invalid. Skip optional spaces and '<', then
+		 * transition to TAG.
+		 *
+		 * TODO: Handle values, such as <section>value</section>.
+		 */
+		case ELEM:
+			if (isspace(*s))
+				break; /* skip spaces */
+			if (*s != '<')
+				return ERROR("invalid token");
+			state = TAG;
+			break; /* skip '<' */
+		/*
+		 * If the current character is '/', skip it and transition to
+		 * END_TAG, otherwise transition to START_TAG.
+		 */
+		case TAG:
+			if (*s != '/') {
+				tag_name = s;
+				state = START_TAG;
+				continue;
+			}
+			if (lua_gettop(L) == top + 1)
+				return ERROR("invalid token");
```

Comment:
TBH, I don't get this condition.

Reply:
Added a comment:

<details><summary>diff</summary>

```diff
diff --git a/src/lua/xml.c b/src/lua/xml.c
index 0eeba3500223..bc5a516b585a 100644
--- a/src/lua/xml.c
+++ b/src/lua/xml.c
@@ -86,8 +86,14 @@ luaT_xml_decode(struct lua_State *L)
 				state = START_TAG;
 				continue;
 			}
-			if (lua_gettop(L) == top + 1)
+			if (lua_gettop(L) == top + 1) {
+				/*
+				 * The stack only contains the resulting
+				 * document table, which means there's no
+				 * start tag matching this end tag.
+				 */
 				return ERROR("invalid token");
+			}
 			tag_name = s + 1;
 			state = END_TAG;
 			break; /* skip '/' */
```

</details>